Add 358 release notes (#8164)
trinodb/trino: docs/src/main/sphinx/release/release-358.md

# Release 358 (XX May 2021)

## General

* Support any query in {doc}`/sql/show-stats`. ({issue}`8026`)
* Improve performance of complex queries involving joins and `TABLESAMPLE`. ({issue}`8094`)
* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)
* Reduce worker graceful shutdown time. ({issue}`8149`)
* Fix query failure when one of join build side's columns is of non-orderable type. ({issue}`7723`)
* Fix failure for queries containing repeated ordinals in `GROUP BY` clause.
  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)
* Fix failure for queries containing repeated expressions in `ORDER BY` clause of an aggregate function.
  Example: `SELECT count(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)  
  
## Security

> * Fix supports for multiple LDAP user bind patterns.  ({issue}`8022`)

Author: @dain, what exactly does this fix? Or is it more like "Add support for ..."?

Reviewer: This was reverted - the correct fix is part of https://github.com/trinodb/trino/pull/8134

Author: Thanks, I'll remove the entry.

## Hive connector

* Add support for reading from Hive views which use `LATERAL VIEW EXPLODE`
  or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCTs`. ({issue}`8120`)
  
## Iceberg connector

> * Fix listing of materialized views in `information_schema.tables` and `system.jdbc.tables` ({issue}`8151`)

Author: @sopel39, what exactly does this fix? Is it a failure? Missing entries? Entries showing up when they shouldn't?

Author: @raunaqmorarka, thoughts?

Reviewer 2: Earlier there was query failure with message `Not an Iceberg table`

> # Release 358 (XX May 2021)

Reviewer 3: Jun

## General

* Support any query in {doc}`/sql/show-stats`. ({issue}`8026`)
* Improve performance of complex queries involving joins and `TABLESAMPLE`. ({issue}`8094`)

> * Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)

Reviewer 4: Should we change to Top-N pushdown since that is the name we use in the docs now? https://trino.io/docs/current/optimizer/pushdown.html#top-n-pushdown

Reviewer 3: I guess "sorted data" here refers to connector-sorted. Should this be under "Hive connector" section instead?

Author: No, we refer to it as ORDER BY ... LIMIT, as that's what users think about when they type their queries.

Reviewer 3: @mosabua topn pushdown is not involved here (no pushdown into the connector takes place in the case study's query)

@martint i unresolved for my other question

Author: > I guess "sorted data" here refers to connector-sorted. Should this be under "Hive connector" section instead?

It's an engine-level improvement. If a connector produces sorted data (and describes its properties to the engine), the engine will be able to take advantage of it. I guess we can add it to both.

Reviewer 5: @mosabua we called it "Top-N pushdown" since we needed a succinct name to refer to that pushdown, and what it means is explained in that documentation section. We don't have a general term "Top-N" defined anywhere.

Reviewer 4: Thanks for clarifying .. I clearly misunderstood.

Author: They can observe this with third party connectors, too, if they are running any. That’s why I think it also belongs in the general section. It’s not specific to the Hive connector.

> * Reduce worker graceful shutdown time. ({issue}`8149`)

Reviewer 4: Reduce graceful shutdown time for workers.

or 

Reduce graceful shutdown time for worker nodes.

Reviewer 3: Worth indicating whether this means reduced guarantees (eg smaller timeout value), or better logic.

> * Fix query failure when one of join build side's columns is of non-orderable type. ({issue}`7723`)

Reviewer 4: Fix query failure when a build side column used in a join uses a non-orderable type.

Author: I'm actually thinking we should rephrase in a way that doesn't say "build side". That's an internal implementation concept, not something users can reason about.

> * Fix failure for queries containing repeated ordinals in `GROUP BY` clause.

Reviewer 4: ```
in the `GROUP BY` clause.
```

or 

```
in an  `GROUP BY` clause.
```

  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)

> * Fix failure for queries containing repeated expressions in `ORDER BY` clause of an aggregate function.

Reviewer 4: ```
in the `ORDER BY` clause
```

or 

```
in an ORDER BY clause
```

>   Example: `SELECT count(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)  

Reviewer 5: Maybe use `array_agg` since that function actually uses `ORDER BY`

> ## JDBC Driver

Reviewer 4: Add empty line below title

> * Remove legacy driver classes `com.facebook.presto.jdbc.PrestoDriver`

Reviewer 4: Is this a breaking change we need to call out more?

> * Remove legacy driver classes `com.facebook.presto.jdbc.PrestoDriver`

Reviewer 5: Nit: missing blank line

>   and `io.prestosql.jdbc.PrestoDriver`. ({issue}`8042`)

Reviewer 5: Maybe we should flip these so the newer `io.prestosql` is first

> * Remove legacy JDBC URL prefix `jdbc:presto:`. ({issue}`8042`)

Reviewer 4: Is this a breaking change we need to call out more?

> * Remove legacy JDBC URL prefix `jdbc:presto:`. ({issue}`8042`)

Reviewer 5: Let's put this first since JDBC URLs are more user facing than driver class names. (I know, these came from me originally...)

## Hive connector

> * Add support for reading from Hive views which use `LATERAL VIEW EXPLODE`

Reviewer 4: views that use

>   or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCTs`. ({issue}`8120`)

Reviewer 3: move "s" outside of backticks and maybe reword to avoid plural.

  
## Iceberg connector

> * Fix failure when listing materialized views in `information_schema.tables` and `system.jdbc.tables` ({issue}`8151`)

Reviewer 3: `system.jdbc.tables` is not really user-facing. `java.sql.DatabaseMetaData#getTables` is

## Memory connector

* Improve performance of certain complex queries involving joins. ({issue}`8095`)

## SPI

> * Remove deprecated ``ConnectorPageSourceProvider#createPageSource`` method overrides. ({issue}`8077`)

Reviewer 5: We should be consistent about this (and fix existing release notes), but I think it looks nicer to write these as `ConnectorPageSourceProvider.createPageSource()`. This better matches The Java Tutorials, etc. I don't feel strongly, but we should pick a style and document it in https://github.com/trinodb/trino/wiki/Release-Notes-Guidelines

> * Add support for casting redirected table columns when source column types don't match. ({issue}`6066`)

Reviewer 6: I think it should be
```
Add support for redirected table columns casting when source column types don't match
```

Reviewer 4: table column casting ??

> * Add support for casting redirected table columns when source column types don't match. ({issue}`6066`)

Reviewer 5: Let's say "redirected table scan" to differentiate from metadata redirection.

> Add support for casting the columns of a redirected table scan when source column types don't match.

> * Add `ConnectorMetadata#redirectTable` to allow the connectors to redirect table scans. ({issue}`7606`)

Reviewer 5: We already have `applyTableScanRedirect()` which is a very different thing, so we should avoid "table scan" here. Let's say 

> to allow connectors to redirect table reads and metadata listings.

> * Add `ConnectorMetadata#streamTableColumns` for streaming columns in a redirection-aware manner. The

Reviewer 6: `streaming columns` -> streaming column metadata`?

# Release 358 (1 Jun 2021)

## General

* Support arbitrary queries in {doc}`/sql/show-stats`. ({issue}`8026`)
* Improve performance of complex queries involving joins and `TABLESAMPLE`. ({issue}`8094`)
* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)
* Reduce graceful shutdown time for worker nodes. ({issue}`8149`)

> * Fix query failure columns of non-orderable types are involved in a join. ({issue}`7723`)

Reviewer 7: Can we enumerate. Or at least give an example what a non-orderable type is?

* Fix failure for queries containing repeated ordinals in a `GROUP BY` clause.
  Example: `SELECT x FROM t GROUP BY 1, 1`. ({issue}`8023`)
* Fix failure for queries containing repeated expressions in the `ORDER BY` clause of an aggregate function.
  Example: `SELECT array_agg(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)

## JDBC Driver

* Remove legacy JDBC URL prefix `jdbc:presto:`. ({issue}`8042`)
* Remove legacy driver classes `io.prestosql.jdbc.PrestoDriver`
  and `com.facebook.presto.jdbc.PrestoDriver`. ({issue}`8042`)

## Hive connector

* Add support for reading from Hive views that use `LATERAL VIEW EXPLODE`
  or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCT`. ({issue}`8120`)
* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)

## Iceberg connector

> * Fix failure when listing materialized views in `information_schema.tables` and `java.sql.DatabaseMetaData.getTables()` ({issue}`8151`)

Reviewer 7: ```suggestion
* Fix failure when listing materialized views using query from `information_schema.tables` or JDBC API call to `java.sql.DatabaseMetaData.getTables()` ({issue}`8151`)
```

## Memory connector

* Improve performance of certain complex queries involving joins. ({issue}`8095`)

## SPI

* Remove deprecated `ConnectorPageSourceProvider.createPageSource()` method overrides. ({issue}`8077`)

> * Add support for casting the columns of a redirected table scan when source column types don't match. ({issue}`6066`)

Reviewer 7: ```suggestion
* Add support for casting columns of a redirected table scan when source column types don't match. ({issue}`6066`)
```

Author: I think "the" is fine in this sentence.

> * Add `ConnectorMetadata#redirectTable` to allow connectors to redirect table reads and metadata listings. ({issue}`7606`)

Reviewer 7: ```suggestion
* Add `ConnectorMetadata#redirectTable` to allow connectors redirect table reads and metadata listings. ({issue}`7606`)
```

Author: "to" is needed: "to allow connectors to <do something>"

* Fix failure for queries containing repeated expressions in the `ORDER BY` clause of an aggregate function.
  Example: `SELECT array_agg(x ORDER BY y, y) FROM (VALUES ('a', 2)) t(x, y)`. ({issue}`8080`)

## JDBC Driver

* Remove legacy JDBC URL prefix `jdbc:presto:`. ({issue}`8042`)
* Remove legacy driver classes `io.prestosql.jdbc.PrestoDriver`
  and `com.facebook.presto.jdbc.PrestoDriver`. ({issue}`8042`)

## Hive connector

* Add support for reading from Hive views that use `LATERAL VIEW EXPLODE`
  or `LATERAL VIEW OUTER EXPLODE` over array of `STRUCT`. ({issue}`8120`)
* Improve performance of `ORDER BY ... LIMIT` queries on sorted data. ({issue}`6634`)

## Iceberg connector

* Fix failure when listing materialized views in `information_schema.tables` or via the 
  `java.sql.DatabaseMetaData.getTables()` JDBC API. ({issue}`8151`)

## Memory connector

* Improve performance of certain complex queries involving joins. ({issue}`8095`)

## SPI

* Remove deprecated `ConnectorPageSourceProvider.createPageSource()` method overrides. ({issue}`8077`)
* Add support for casting the columns of a redirected table scan when source column types don't match. ({issue}`6066`)

> * Add `ConnectorMetadata#redirectTable` to allow connectors to redirect table reads and metadata listings. ({issue}`7606`)

Reviewer 5: `ConnectorMetadata.redirectTable()`

> * Add `ConnectorMetadata#streamTableColumns` for streaming column metadata in a redirection-aware manner. The

Reviewer 5: `ConnectorMetadata.streamTableColumns()`

>   alternate method for column listing `ConnectorMetadata#listTableColumns` is now deprecated. ({issue}`7606`)

Reviewer 5: `ConnectorMetadata.listTableColumns()`
